### PR TITLE
Rewrite of TSRC (TypeScript Runtime Checks)

### PR DESCRIPTION
**Pull request by hgossler**
## Overview
This pull request introduces a comprehensive rewrite of the TypeScript Run-time Checks (TSRC) tool, previously available under `@omegadot/tsrc`. The new implementation brings significant improvements in speed and offers a better DX, as the developer does not need to specify which types to generate checks for up front.

All imports to the old type checks have been replaced.

## Implementation Details
The TSRC tool operates in two main steps:
1. **Scanning for Type Declarations**: Initially, it scans the project's source files for type declarations eligible for runtime type checks, generating ambient module declarations as hints for the IDE.
2. **Generating Runtime Check Implementations**: Subsequently, it generates the actual runtime check implementations by scanning for imports to the previously generated ambient module declarations and utilizing Ajv for the validation functions.

### CODE REVIEW

File: scripts/tsrc/TSRC.ts [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+			// Detect which declarations have been used
+			if (ts.isImportDeclaration(node)) {
+				// We only care about imports from the special virtual module
+				// Get the module specifier without quotes
+				const module = node.moduleSpecifier.getFullText(sourceFile).trim().slice(1, -1);
+				// Append a slash to the virtual module name to avoid matching modules with a coincidental
+				// common prefix
+				if (!module.startsWith(`${this.virtualModuleName}/`)) return;
+				// Omit imports declarations that don't import anything
+				if (!node.importClause) return;
```
Comment (joeried) [inline scripts/tsrc/TSRC.ts:176]: If this call is faster than `node.moduleSpecifier.getFullText()` we should consider swapping the order of these returns.

File: scripts/__tests__/tsrc.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	});
+});
+
+export interface ITest {
+	a: string;
+	b: number;
+}
+
+export type Intersection = ITest & { c: boolean };
+export type Union = ITest | { c: boolean };
```
Comment (joeried) [inline scripts/__tests__/tsrc.spec.ts:80]: This test file would be easier to read if the types to which the test refers were declared at the beginning

File: scripts/tsrc/TSRC.ts [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+	 * @param virtualModuleName - Directory relative to the project directory where generated files are written to. Specifying a path within `node_modules` breaks auto-imports in the IDE.
+	 * @param ignore - List of paths relative to the project directory to ignore.
+	 */
+	constructor(
+		private readonly projectDirectory: string,
+		virtualModuleName: string,
+		ignore: string[] = []
+	) {
+		// Interpret relative paths as being relative to projectDirectory.
+		// Calling `resolve` has no effect if `ignorePath` is absolute, so it's safe to call it unconditionally.
```
[the remark this answers is not in the document]
Comment (hgossler) [reply scripts/tsrc/TSRC.ts]: Fixed

File: scripts/tsrc/TSRC.ts [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+ * to be consumed directly by the user.
+ */
+export class TSRC {
+	/**
+	 * The absolute path to the directory where generated files are written to.
+	 */
+	private readonly outDirectory: string;
+
+	/**
+	 * List of absolute paths of files to ignore.
```
[the remark this answers is not in the document]
Comment (hgossler) [reply scripts/tsrc/TSRC.ts]: Fixed

File: scripts/tsrc/TSRC.ts (hunk 1 above)
Comment (hgossler) [reply scripts/tsrc/TSRC.ts]: Fixed

File: scripts/__tests__/tsrc.spec.ts (hunk 2 above)
Comment (hgossler) [reply scripts/__tests__/tsrc.spec.ts]: Fixed

**Pull request merged**